File: plugins/plugin-secrets-manager/python/elizaos_plugin_secrets_manager/validation.py
"""
Secret Validation module.

Provides validation strategies for common API keys and secret formats.
"""

import re
import logging
from typing import Optional, Dict, Callable, Awaitable
from dataclasses import dataclass


logger = logging.getLogger(__name__)


@dataclass
class ValidationResult:
    """Result of secret validation."""
    is_valid: bool
    error: Optional[str] = None


# Type alias for validators
ValidationStrategy = Callable[[str, str], Awaitable[ValidationResult]]


# Registry for validation strategies
_validation_strategies: Dict[str, ValidationStrategy] = {}


def register_validator(name: str, validator: ValidationStrategy) -> None:
    """Register a validation strategy."""
    _validation_strategies[name] = validator


def get_validator(name: str) -> Optional[ValidationStrategy]:
    """Get a validation strategy by name."""
    return _validation_strategies.get(name)


async def validate_secret(
    key: str,
    value: str,
    method: Optional[str] = None,
) -> ValidationResult:
    """Validate a secret value."""
    if not method or method == "none":
        return ValidationResult(is_valid=True)
    
    # Try to get registered validator
    validator = get_validator(method)
    if validator:
        return await validator(key, value)
    
    # Fall back to built-in strategies
    if method in BUILT_IN_VALIDATORS:
        return await BUILT_IN_VALIDATORS[method](key, value)
    
    # Auto-detect based on key name
    if method == "auto":
        inferred = infer_validation_strategy(key)
        if inferred and inferred in BUILT_IN_VALIDATORS:
            return await BUILT_IN_VALIDATORS[inferred](key, value)
    
    # Unknown method - pass
    logger.warning(f"Unknown validation method: {method}")
    return ValidationResult(is_valid=True)


def infer_validation_strategy(key: str) -> Optional[str]:
    """Infer validation strategy from key name."""
    key_lower = key.lower()
    
    if "openai" in key_lower:
        return "openai"
    if "anthropic" in key_lower or "claude" in key_lower:
        return "anthropic"
    if "groq" in key_lower:
        return "groq"
    if "google" in key_lower or "gemini" in key_lower:
        return "google"
    if "mistral" in key_lower:
        return "mistral"
    if "cohere" in key_lower:
        return "cohere"
    if "url" in key_lower or "endpoint" in key_lower:
        return "url"
    if "discord" in key_lower:
        return "discord"
    if "telegram" in key_lower:
        return "telegram"
    if "github" in key_lower:
        return "github"
    
    return None


# ============================================================================
# Built-in Validation Strategies
# ============================================================================

async def validate_openai(key: str, value: str) -> ValidationResult:
    """Validate OpenAI API key format."""
    if not value:
        return ValidationResult(is_valid=False, error="Empty value")
    
    # OpenAI keys start with sk- and are typically 51 chars
    if not value.startswith("sk-"):
        return ValidationResult(is_valid=False, error="OpenAI keys must start with 'sk-'")
    
    if len(value) < 20:
        return ValidationResult(is_valid=False, error="OpenAI key too short")
    
    # Check for valid characters
    if not re.match(r"^sk-[a-zA-Z0-9_-]+$", value):
        return ValidationResult(is_valid=False, error="Invalid characters in OpenAI key")
    
    return ValidationResult(is_valid=True)


async def validate_anthropic(key: str, value: str) -> ValidationResult:
    """Validate Anthropic API key format."""
    if not value:
        return ValidationResult(is_valid=False, error="Empty value")
    
    # Anthropic keys start with sk-ant-
    if not value.startswith("sk-ant-"):
        return ValidationResult(is_valid=False, error="Anthropic keys must start with 'sk-ant-'")
    
    if len(value) < 20:
        return ValidationResult(is_valid=False, error="Anthropic key too short")
    
    return ValidationResult(is_valid=True)


async def validate_groq(key: str, value: str) -> ValidationResult:
    """Validate Groq API key format."""
    if not value:
        return ValidationResult(is_valid=False, error="Empty value")
    
    # Groq keys start with gsk_
    if not value.startswith("gsk_"):
        return ValidationResult(is_valid=False, error="Groq keys must start with 'gsk_'")
    
    if len(value) < 20:
        return ValidationResult(is_valid=False, error="Groq key too short")
    
    return ValidationResult(is_valid=True)


async def validate_google(key: str, value: str) -> ValidationResult:
    """Validate Google/Gemini API key format."""
    if not value:
        return ValidationResult(is_valid=False, error="Empty value")
    
    # Google API keys are typically 39 chars and alphanumeric
    if len(value) < 30:
        return ValidationResult(is_valid=False, error="Google API key too short")
    
    if not re.match(r"^[a-zA-Z0-9_-]+$", value):
        return ValidationResult(is_valid=False, error="Invalid characters in Google API key")
    
    return ValidationResult(is_valid=True)


async def validate_mistral(key: str, value: str) -> ValidationResult:
    """Validate Mistral API key format."""
    if not value:
        return ValidationResult(is_valid=False, error="Empty value")
    
    if len(value) < 20:
        return ValidationResult(is_valid=False, error="Mistral key too short")
    
    return ValidationResult(is_valid=True)


async def validate_cohere(key: str, value: str) -> ValidationResult:
    """Validate Cohere API key format."""
    if not value:
        return ValidationResult(is_valid=False, error="Empty value")
    
    if len(value) < 20:
        return ValidationResult(is_valid=False, error="Cohere key too short")
    
    # Cohere keys are typically alphanumeric
    if not re.match(r"^[a-zA-Z0-9_-]+$", value):
        return ValidationResult(is_valid=False, error="Invalid characters in Cohere key")
    
    return ValidationResult(is_valid=True)


async def validate_url(key: str, value: str) -> ValidationResult:
    """Validate URL format."""
    if not value:
        return ValidationResult(is_valid=False, error="Empty value")
    
    # Basic URL regex
    url_pattern = r"^https?://[^\s/$.?#].[^\s]*$"
    if not re.match(url_pattern, value, re.IGNORECASE):
        return ValidationResult(is_valid=False, error="Invalid URL format")
    
    return ValidationResult(is_valid=True)


async def validate_discord(key: str, value: str) -> ValidationResult:
    """Validate Discord bot token format."""
    if not value:
        return ValidationResult(is_valid=False, error="Empty value")
    
    # Discord tokens have a specific format with dots
    parts = value.split(".")
    if len(parts) != 3:
        return ValidationResult(is_valid=False, error="Discord token must have 3 parts separated by dots")
    
    if len(value) < 50:
        return ValidationResult(is_valid=False, error="Discord token too short")
    
    return ValidationResult(is_valid=True)


async def validate_telegram(key: str, value: str) -> ValidationResult:
    """Validate Telegram bot token format."""
    if not value:
        return ValidationResult(is_valid=False, error="Empty value")
    
    # Telegram tokens have format: 123456:ABC-DEF1234ghIkl-zyx57W2v1u123ew11
    pattern = r"^\d+:[A-Za-z0-9_-]+$"
    if not re.match(pattern, value):
        return ValidationResult(is_valid=False, error="Invalid Telegram token format")
    
    return ValidationResult(is_valid=True)


async def validate_github(key: str, value: str) -> ValidationResult:
    """Validate GitHub token format."""
    if not value:
        return ValidationResult(is_valid=False, error="Empty value")
    
    # GitHub tokens start with ghp_, gho_, ghu_, ghs_, or ghr_
    # Classic tokens start with ghp_, fine-grained with github_pat_
    valid_prefixes = ("ghp_", "gho_", "ghu_", "ghs_", "ghr_", "github_pat_")
    if not value.startswith(valid_prefixes):
        return ValidationResult(
            is_valid=False,
            error="GitHub token must start with ghp_, gho_, ghu_, ghs_, ghr_, or github_pat_"
        )
    
    if len(value) < 30:
        return ValidationResult(is_valid=False, error="GitHub token too short")
    
    return ValidationResult(is_valid=True)


async def validate_non_empty(key: str, value: str) -> ValidationResult:
    """Simple non-empty validation."""
    if not value or not value.strip():
        return ValidationResult(is_valid=False, error="Value cannot be empty")
    return ValidationResult(is_valid=True)


async def validate_min_length(min_len: int):
    """Create a minimum length validator."""
    async def validator(key: str, value: str) -> ValidationResult:
        if not value:
            return ValidationResult(is_valid=False, error="Empty value")
        if len(value) < min_len:
            return ValidationResult(
                is_valid=False,
                error=f"Value must be at least {min_len} characters"
            )
        return ValidationResult(is_valid=True)
    return validator


# Built-in validators registry
BUILT_IN_VALIDATORS: Dict[str, ValidationStrategy] = {
    "openai": validate_openai,
    "anthropic": validate_anthropic,
    "groq": validate_groq,
    "google": validate_google,
    "gemini": validate_google,  # Alias
    "mistral": validate_mistral,
    "cohere": validate_cohere,
    "url": validate_url,
    "discord": validate_discord,
    "telegram": validate_telegram,
    "github": validate_github,
    "non_empty": validate_non_empty,
}

File: plugins/plugin-secrets-manager/rust/src/validation.rs
//! Secret validation module.
//!
//! Provides validation strategies for common API keys and secret formats.

use once_cell::sync::Lazy;
use regex::Regex;
use std::collections::HashMap;
use std::sync::RwLock;

/// Result of secret validation.
#[derive(Debug, Clone)]
pub struct ValidationResult {
    /// Whether the value is valid.
    pub is_valid: bool,
    /// Error message if invalid.
    pub error: Option<String>,
}

impl ValidationResult {
    /// Create a valid result.
    pub fn valid() -> Self {
        Self {
            is_valid: true,
            error: None,
        }
    }

    /// Create an invalid result with an error message.
    pub fn invalid(error: impl Into<String>) -> Self {
        Self {
            is_valid: false,
            error: Some(error.into()),
        }
    }
}

/// Type alias for validation functions.
pub type ValidationFn = fn(&str, &str) -> ValidationResult;

/// Registry of custom validation strategies.
static CUSTOM_VALIDATORS: Lazy<RwLock<HashMap<String, ValidationFn>>> =
    Lazy::new(|| RwLock::new(HashMap::new()));

/// Register a custom validation strategy.
pub fn register_validator(name: &str, validator: ValidationFn) {
    let mut validators = CUSTOM_VALIDATORS.write().unwrap();
    validators.insert(name.to_string(), validator);
}

/// Get a custom validation strategy.
pub fn get_validator(name: &str) -> Option<ValidationFn> {
    let validators = CUSTOM_VALIDATORS.read().unwrap();
    validators.get(name).copied()
}

/// Validate a secret value.
pub fn validate_secret(key: &str, value: &str, method: Option<&str>) -> ValidationResult {
    let method = match method {
        Some(m) if !m.is_empty() && m != "none" => m,
        _ => return ValidationResult::valid(),
    };

    // Check custom validators first
    if let Some(validator) = get_validator(method) {
        return validator(key, value);
    }

    // Built-in validators
    match method {
        "openai" => validate_openai(key, value),
        "anthropic" => validate_anthropic(key, value),
        "groq" => validate_groq(key, value),
        "google" | "gemini" => validate_google(key, value),
        "mistral" => validate_mistral(key, value),
        "cohere" => validate_cohere(key, value),
        "url" => validate_url(key, value),
        "discord" => validate_discord(key, value),
        "telegram" => validate_telegram(key, value),
        "github" => validate_github(key, value),
        "non_empty" => validate_non_empty(key, value),
        "auto" => {
            if let Some(inferred) = infer_validation_strategy(key) {
                validate_secret(key, value, Some(&inferred))
            } else {
                ValidationResult::valid()
            }
        }
        _ => ValidationResult::valid(),
    }
}

/// Infer validation strategy from key name.
pub fn infer_validation_strategy(key: &str) -> Option<String> {
    let key_lower = key.to_lowercase();

    if key_lower.contains("openai") {
        return Some("openai".to_string());
    }
    if key_lower.contains("anthropic") || key_lower.contains("claude") {
        return Some("anthropic".to_string());
    }
    if key_lower.contains("groq") {
        return Some("groq".to_string());
    }
    if key_lower.contains("google") || key_lower.contains("gemini") {
        return Some("google".to_string());
    }
    if key_lower.contains("mistral") {
        return Some("mistral".to_string());
    }
    if key_lower.contains("cohere") {
        return Some("cohere".to_string());
    }
    if key_lower.contains("url") || key_lower.contains("endpoint") {
        return Some("url".to_string());
    }
    if key_lower.contains("discord") {
        return Some("discord".to_string());
    }
    if key_lower.contains("telegram") {
        return Some("telegram".to_string());
    }
    if key_lower.contains("github") {
        return Some("github".to_string());
    }

    None
}

// ============================================================================
// Built-in Validators
// ============================================================================

/// Validate OpenAI API key format.
pub fn validate_openai(_key: &str, value: &str) -> ValidationResult {
    if value.is_empty() {
        return ValidationResult::invalid("Empty value");
    }

    if !value.starts_with("sk-") {
        return ValidationResult::invalid("OpenAI keys must start with 'sk-'");
    }

    if value.len() < 20 {
        return ValidationResult::invalid("OpenAI key too short");
    }

    static PATTERN: Lazy<Regex> = Lazy::new(|| Regex::new(r"^sk-[a-zA-Z0-9_-]+$").unwrap());
    if !PATTERN.is_match(value) {
        return ValidationResult::invalid("Invalid characters in OpenAI key");
    }

    ValidationResult::valid()
}

/// Validate Anthropic API key format.
pub fn validate_anthropic(_key: &str, value: &str) -> ValidationResult {
    if value.is_empty() {
        return ValidationResult::invalid("Empty value");
    }

    if !value.starts_with("sk-ant-") {
        return ValidationResult::invalid("Anthropic keys must start with 'sk-ant-'");
    }

    if value.len() < 20 {
        return ValidationResult::invalid("Anthropic key too short");
    }

    ValidationResult::valid()
}

/// Validate Groq API key format.
pub fn validate_groq(_key: &str, value: &str) -> ValidationResult {
    if value.is_empty() {
        return ValidationResult::invalid("Empty value");
    }

    if !value.starts_with("gsk_") {
        return ValidationResult::invalid("Groq keys must start with 'gsk_'");
    }

    if value.len() < 20 {
        return ValidationResult::invalid("Groq key too short");
    }

    ValidationResult::valid()
}

/// Validate Google/Gemini API key format.
pub fn validate_google(_key: &str, value: &str) -> ValidationResult {
    if value.is_empty() {
        return ValidationResult::invalid("Empty value");
    }

    if value.len() < 30 {
        return ValidationResult::invalid("Google API key too short");
    }

    static PATTERN: Lazy<Regex> = Lazy::new(|| Regex::new(r"^[a-zA-Z0-9_-]+$").unwrap());
    if !PATTERN.is_match(value) {
        return ValidationResult::invalid("Invalid characters in Google API key");
    }

    ValidationResult::valid()
}

/// Validate Mistral API key format.
pub fn validate_mistral(_key: &str, value: &str) -> ValidationResult {
    if value.is_empty() {
        return ValidationResult::invalid("Empty value");
    }

    if value.len() < 20 {
        return ValidationResult::invalid("Mistral key too short");
    }

    ValidationResult::valid()
}

/// Validate Cohere API key format.
pub fn validate_cohere(_key: &str, value: &str) -> ValidationResult {
    if value.is_empty() {
        return ValidationResult::invalid("Empty value");
    }

    if value.len() < 20 {
        return ValidationResult::invalid("Cohere key too short");
    }

    static PATTERN: Lazy<Regex> = Lazy::new(|| Regex::new(r"^[a-zA-Z0-9_-]+$").unwrap());
    if !PATTERN.is_match(value) {
        return ValidationResult::invalid("Invalid characters in Cohere key");
    }

    ValidationResult::valid()
}

/// Validate URL format.
pub fn validate_url(_key: &str, value: &str) -> ValidationResult {
    if value.is_empty() {
        return ValidationResult::invalid("Empty value");
    }

    static PATTERN: Lazy<Regex> =
        Lazy::new(|| Regex::new(r"(?i)^https?://[^\s/$.?#].[^\s]*$").unwrap());
    if !PATTERN.is_match(value) {
        return ValidationResult::invalid("Invalid URL format");
    }

    ValidationResult::valid()
}

/// Validate Discord bot token format.
pub fn validate_discord(_key: &str, value: &str) -> ValidationResult {
    if value.is_empty() {
        return ValidationResult::invalid("Empty value");
    }

    let parts: Vec<&str> = value.split('.').collect();
    if parts.len() != 3 {
        return ValidationResult::invalid("Discord token must have 3 parts separated by dots");
    }

    if value.len() < 50 {
        return ValidationResult::invalid("Discord token too short");
    }

    ValidationResult::valid()
}

/// Validate Telegram bot token format.
pub fn validate_telegram(_key: &str, value: &str) -> ValidationResult {
    if value.is_empty() {
        return ValidationResult::invalid("Empty value");
    }

    static PATTERN: Lazy<Regex> = Lazy::new(|| Regex::new(r"^\d+:[A-Za-z0-9_-]+$").unwrap());
    if !PATTERN.is_match(value) {
        return ValidationResult::invalid("Invalid Telegram token format");
    }

    ValidationResult::valid()
}

/// Validate GitHub token format.
pub fn validate_github(_key: &str, value: &str) -> ValidationResult {
    if value.is_empty() {
        return ValidationResult::invalid("Empty value");
    }

    let valid_prefixes = ["ghp_", "gho_", "ghu_", "ghs_", "ghr_", "github_pat_"];
    if !valid_prefixes.iter().any(|p| value.starts_with(p)) {
        return ValidationResult::invalid(
            "GitHub token must start with ghp_, gho_, ghu_, ghs_, ghr_, or github_pat_",
        );
    }

    if value.len() < 30 {
        return ValidationResult::invalid("GitHub token too short");
    }

    ValidationResult::valid()
}

/// Simple non-empty validation.
pub fn validate_non_empty(_key: &str, value: &str) -> ValidationResult {
    if value.is_empty() || value.trim().is_empty() {
        return ValidationResult::invalid("Value cannot be empty");
    }
    ValidationResult::valid()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_validate_openai() {
        assert!(validate_openai("", "sk-abc123def456ghi789jkl").is_valid);
        assert!(!validate_openai("", "").is_valid);
        assert!(!validate_openai("", "invalid-key").is_valid);
        assert!(!validate_openai("", "sk-abc").is_valid); // Too short
    }

    #[test]
    fn test_validate_anthropic() {
        assert!(validate_anthropic("", "sk-ant-abc123def456ghi789").is_valid);
        assert!(!validate_anthropic("", "").is_valid);
        assert!(!validate_anthropic("", "sk-abc123").is_valid);
    }

    #[test]
    fn test_validate_groq() {
        assert!(validate_groq("", "gsk_abc123def456ghi789jkl").is_valid);
        assert!(!validate_groq("", "").is_valid);
        assert!(!validate_groq("", "invalid").is_valid);
    }

    #[test]
    fn test_validate_url() {
        assert!(validate_url("", "https://example.com").is_valid);
        assert!(validate_url("", "http://localhost:8080/api").is_valid);
        assert!(!validate_url("", "").is_valid);
        assert!(!validate_url("", "not-a-url").is_valid);
    }

    #[test]
    fn test_validate_discord() {
        assert!(validate_discord("", "MTExMTExMTExMTExMTExMTEx.XXXXXX.XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX").is_valid);
        assert!(!validate_discord("", "invalid").is_valid);
    }

    #[test]
    fn test_validate_telegram() {
        assert!(validate_telegram("", "123456789:ABCdefGHIjklMNOpqrSTUvwxYZ").is_valid);
        assert!(!validate_telegram("", "invalid").is_valid);
    }

    #[test]
    fn test_validate_github() {
        assert!(validate_github("", "ghp_xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx").is_valid);
        assert!(validate_github("", "github_pat_xxxxxxxxxxxxxxxxxxxx").is_valid);
        assert!(!validate_github("", "invalid").is_valid);
    }

    #[test]
    fn test_infer_validation_strategy() {
        assert_eq!(infer_validation_strategy("OPENAI_API_KEY"), Some("openai".to_string()));
        assert_eq!(infer_validation_strategy("ANTHROPIC_API_KEY"), Some("anthropic".to_string()));
        assert_eq!(infer_validation_strategy("GROQ_API_KEY"), Some("groq".to_string()));
        assert_eq!(infer_validation_strategy("DATABASE_URL"), Some("url".to_string()));
        assert_eq!(infer_validation_strategy("RANDOM_KEY"), None);
    }

    #[test]
    fn test_validate_secret_auto() {
        // Auto should infer OpenAI validation
        let result = validate_secret("OPENAI_API_KEY", "sk-abc123def456ghi789jkl", Some("auto"));
        assert!(result.is_valid);

        let result = validate_secret("OPENAI_API_KEY", "invalid", Some("auto"));
        assert!(!result.is_valid);
    }

    #[test]
    fn test_custom_validator() {
        fn custom_validator(_key: &str, value: &str) -> ValidationResult {
            if value.starts_with("custom_") {
                ValidationResult::valid()
            } else {
                ValidationResult::invalid("Must start with custom_")
            }
        }

        register_validator("custom", custom_validator);

        let result = validate_secret("key", "custom_value", Some("custom"));
        assert!(result.is_valid);

        let result = validate_secret("key", "invalid_value", Some("custom"));
        assert!(!result.is_valid);
    }
}
